device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java

```diff
@@ -178,61 +158,47 @@ public void addDeviceOperationSession(DeviceClientConfig deviceClientConfig) thr
-
         this.savedException = null;
 
-        this.sasTokenRenewalHandler = new AmqpSasTokenRenewalHandler(this.amqpsSessionManager, this.deviceClientConfig);
-
-
-        // Codes_SRS_AMQPSIOTHUBCONNECTION_15_007: [If the AMQPS connection is already open, the function shall do nothing.]
         if (this.state == IotHubConnectionStatus.DISCONNECTED)
         {
-            if (deviceClientConfigs.size() > 1)
+            for (DeviceClientConfig clientConfig : deviceClientConfigs)
             {
-                deviceClientConfigs.remove();
-                while (!deviceClientConfigs.isEmpty())
-                {
-                    this.addDeviceOperationSession(deviceClientConfigs.remove());
-                }
+                this.addDeviceSession(clientConfig, this, false);
             }
 
-            initializeStateLatches(deviceClientConfigs.size());
+            initializeStateLatches();
 
-            // Codes_SRS_AMQPSIOTHUBCONNECTION_15_010: [The function shall wait for the reactor to be ready and for
-            // enough link credit to become available.]
             try
             {
-                // Codes_SRS_AMQPSIOTHUBCONNECTION_15_009: [The function shall trigger the Reactor (Proton) to begin running.]
                 this.openAsync();
 
-                // Codes_SRS_AMQPSIOTHUBCONNECTION_12_059: [The function shall call await on open latch.]
                 this.log.trace("Waiting for authentication links to open...");
-                boolean authenticationLinksOpenTimedOut = !this.authenticationLinkOpenLatch.await(MAX_WAIT_TO_OPEN_CBS_LINKS, TimeUnit.MILLISECONDS);
+                boolean authenticationSessionOpenedTimedOut = !this.authenticationSessionOpenedLatch.await(MAX_WAIT_TO_OPEN_CBS_LINKS, TimeUnit.MILLISECONDS);
 
                 if (this.savedException != null)
                 {
                     throw this.savedException;
                 }
 
-                if (authenticationLinksOpenTimedOut)
+                if (authenticationSessionOpenedTimedOut)
                 {
-                    closeConnectionWithException("Timed out waiting for authentication links to open", true);
+                    closeConnectionWithException("Timed out waiting for authentication session to open", true);
                 }
 
-                this.log.trace("Waiting for worker links to open...");
-                boolean workerLinksOpenTimedOut = !this.workerLinksOpenLatch.await(MAX_WAIT_TO_OPEN_WORKER_LINKS, TimeUnit.MILLISECONDS);
+                this.log.trace("Waiting for device sessions to open...");
+                boolean deviceSessionsOpenTimedOut = !this.deviceSessionsOpenedLatch.await(MAX_WAIT_TO_OPEN_WORKER_LINKS, TimeUnit.MILLISECONDS);
 
                 if (this.savedException != null)
                 {
                     throw this.savedException;
                 }
 
-                if (workerLinksOpenTimedOut)
+                if (deviceSessionsOpenTimedOut)
```

Comment:
what happens if some sessions open, and some time out - how are they closed? 
do we rely on client connection closure -> remote connection closure -> remote session/ links closure -> client session/ link closure?

Reply:
If some links don't open, then we do shut down the connection, all sessions in that connection, and all links in those sessions. This is handled by the retry logic in the IotHubTransport layer where the amqp layer is closed, then re-opened